### ISSUE TITLE
CSP blocks event browser resources from loading

### ISSUE THREAD

**Issue by UsAndRufus** [RailsEventStore/rails_event_store#1346]
# Context

We have quite a strict CSP for security reasons. We enable it in dev too to make sure we don't deploy code that doesn't work due to CSP issues. Part of this CSP configuration is using `nonced: true` on all inline JavaScript. This seems pretty standard for modern frontend security.

The issue is that the RES Browser uses an un-nonced inline JavsaScript tag. This means the app doesn't load.

![Screenshot 2022-06-17 at 14 33 28](https://user-images.githubusercontent.com/4115329/174308653-836f2f2b-3be1-41e9-8f58-22726c615a26.png)

We have other apps with significant frontend JS (eg BetterErrors) work fine. I'm not sure what they're doing differently.

# Workaround
My workaround is to build a simple event view using our admin tools for now.

# Potential Solution

I _think_ the issue would be fixed if the main entrypoint to Elm wasn't inlined but rather was pulled from the server like the other JS files. Rails seems to have no issue with the other sources listed in the page as I believe everything served from the host is considered fine.

![Screenshot 2022-06-17 at 14 36 47](https://user-images.githubusercontent.com/4115329/174309221-0782d373-e630-418a-a9c3-a2ace8454c06.png)

IE it's that second script tag that is the issue

# Environment

We are on Rails 7.x, recently upgraded from Rails 6.x so our app is very 6-ish.

## CSP Policy
From when run locally, hence the references to `localhost` and `.test`
```
Content-Security-Policy: base-uri 'none'; default-src 'none'; frame-src 'self' https:; frame-ancestors 'self' https: http:; font-src 'self' https://*.typekit.net https://fonts.gstatic.com; img-src 'self' https: data: https://provenance.test; manifest-src 'self' https://provenance.test; media-src 'self' https: https://provenance.test; object-src 'none'; style-src 'self' https: https://provenance.test 'nonce-GdEBfK4Gs0N3yzPuwonxPw=='; script-src 'self' https://provenance.test https://www.google-analytics.com https://www.googletagmanager.com https://code.jquery.com https://cdnjs.cloudflare.com https://maxcdn.bootstrapcdn.com 'nonce-GdEBfK4Gs0N3yzPuwonxPw=='; child-src 'none'; worker-src 'none'; connect-src 'self' https: https://provenance.test https://upload.filestackapi.com http://provenance.test http://provenance.test:3000 http://localhost:8545 http://localhost:5001 https://ipfs.infura.io:5001 https://ipfs-cluster.provenance.org https://kovan.infura.io https://mainnet.infura.io wss://localhost:3035 https://localhost:3035
```

# Changelog

_Edit 14:30_
Worked out the issue was the nonce rather than source list, updated issue accordingly.

**Comment by UsAndRufus** [comment]
Happy to open a PR for docs on the CSP issue, ie describe how to add RES Browser to the CSP. But would be nice if we didn't have this issue to begin with, a la BetterErrors

**Comment by pawelpacana** [comment]
> I think the issue would be fixed if the main entrypoint to Elm wasn't inlined

That has changed in current [2.4.0 release](https://github.com/RailsEventStore/rails_event_store/releases/tag/v2.4.0).

Currently the script tags look like this, as seen on https://ecommerce.arkademy.dev/res:

```html
<!DOCTYPE html>
<html>
  <head>
    <title>RubyEventStore::Browser</title>
    <meta
      name="ruby-event-store-browser-settings"
      content='{"rootUrl":"https://ecommerce.arkademy.dev/res","apiUrl":"https://ecommerce.arkademy.dev/res/api","resVersion":"2.4.1"}'
    />
  </head>
  <body>
    <script type="text/javascript" src="/res/ruby_event_store_browser.js"></script>
    <script type="text/javascript" src="/res/bootstrap.js"></script>
  </body>
</html>

```
Would that still be an issue?

**Comment by pawelpacana** [comment]
For reference: https://github.com/RailsEventStore/rails_event_store/issues/1062

**Comment by UsAndRufus** [comment]
Brilliant, thanks! I'll upgrade

**Comment by UsAndRufus** [comment]
Apologies for not checking my Gem version, should have been first port of call. The upgrade sort of works! The JS is allowed and loads the events, but CSS is blocked because it's all inline and we have a strict `style-src` in our CSP (included above).

![Screenshot 2022-06-21 at 12 01 58](https://user-images.githubusercontent.com/4115329/174784715-ec5738e9-8248-4c57-b545-3904b02a48ce.png)

![Screenshot 2022-06-21 at 12 02 09](https://user-images.githubusercontent.com/4115329/174784737-cb24f839-a62e-4c10-9f39-08167a8a28a4.png)

It's not the end of the world though, I can browse the events still. Just takes me back to 1999 😅

**Comment by pawelpacana** [comment]
I'll look into CSS issue (coming from https://github.com/rtfeldman/elm-css)

**Comment by pawelpacana** [comment]
https://github.com/rtfeldman/elm-css/issues/569

**Comment by pawelpacana** [comment]
Elm-css and nonces turned out to be a dead end. I've brought back external stylesheet and that would definitely work with inline-styles disabled via CSP.

**Comment by pawelpacana** [comment]
Fixed and released in [v2.5.0](https://github.com/RailsEventStore/rails_event_store/releases/tag/v2.5.0)

**Comment by UsAndRufus** [comment]
Thanks so much!
